centreon/src/Core/Common/Infrastructure/Repository/ApiCallIterator.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+/*
+ * Copyright 2005 - 2023 Centreon (https://www.centreon.com/)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+declare(strict_types = 1);
+
+namespace Core\Common\Infrastructure\Repository;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Contracts\HttpClient\Exception\ClientExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\RedirectionExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\ServerExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @template T
+ *
+ * @implements \IteratorAggregate<int, T>&\Countable
+ */
+class ApiCallIterator implements \IteratorAggregate, \Countable
+{
+    use ApiResponseTrait;
+
+    private int $nbrElements = 0;
+
+    /** @var list<T> */
+    private array $entitiesCache = [];
+
+    private bool $isCachedCreated = false;
+
+    /**
+     * @param HttpClientInterface $httpClient
+     * @param string $url
+     * @param array<string, mixed> $options
+     * @param int $maxItemsByRequest
+     * @param \Closure $entityFactory
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        private readonly HttpClientInterface $httpClient,
+        private readonly string $url,
+        private readonly array $options,
+        private readonly int $maxItemsByRequest,
+        private readonly \Closure $entityFactory,
```

Comment:
I would a used an interface with a `public function __invoke(array $data): object` instead of `Closure` in order to enforce the closure signature.

Reply:
Yes but the first use will be as a closure.
Later, if the need arises, we'll modify the signature.

Reply:
An object instance with an `__invoke` method **is** a `Closure` but specialized.
It guarantees the `Closure` signature.

Because passing any generic `Closure` can lead to problem if this generic `Closure` does not return an object for the purpose of entity factory.

But anyway, that's a design decision up to you for the later usage.
:point_right: Fell free to resolve my comment if you don't think it is worth.